var/spack/repos/builtin/packages/py-chainer/package.py

```diff
@@ -0,0 +1,42 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+# ----------------------------------------------------------------------------
+
+from spack import *
+
+
+class PyChainer(PythonPackage):
+    """
+    Chainer is a Python-based deep learning framework aiming at flexibility.
+
+    It provides automatic differentiation APIs based on the define-by-run
+    approach (a.k.a. dynamic computational graphs) as well as object-oriented
+    high-level APIs to build and train neural networks.
+    It also supports CUDA/cuDNN using CuPy for high performance training
+    and inference.
+    """
+
+    homepage = "https://chainer.org/"
+    url      = "https://github.com/chainer/chainer/archive/v7.2.0.tar.gz"
+
+    maintainers = ['adamjstewart']
+
+    version('7.2.0', sha256='6e2fba648cc5b8a5421e494385b76fe5ec154f1028a1c5908557f5d16c04f0b3')
+    version('6.7.0', sha256='87cb3378a35e7c5c695028ec91d58dc062356bc91412384ea939d71374610389')
+
+    depends_on('python@3.5.1:', when='@7:', type=('build', 'run'))
+    depends_on('py-setuptools', type=('build', 'run'))
+    depends_on('py-numpy@1.9:', type=('build', 'run'))
+    depends_on('py-six@1.9.0:', type=('build', 'run'))
+    depends_on('py-typing-extensions', type=('build', 'run'))
+    depends_on(
+        'py-typing-extensions@:3.6.6', when='@:6', type=('build', 'run')
+    )
+    depends_on('py-filelock', type=('build', 'run'))
+    depends_on('py-protobuf@3:', type=('build', 'run'))
+    depends_on(
+        'py-typing@:3.6.6', when='@:6 ^python@:3.6', type=('build', 'run')
```

Comment:
```suggestion
        'py-typing@:3.6.6', when='@:6', type=('build', 'run')
```
The dependency in the package is unconditional, even if it is built into Python at that version.

Reply:
But I install py-chainer@6.7.0 on python 3.7.7, `import chainer `is failed as following:
<pre>
$ python
Python 3.7.7 (default, Apr 23 2020, 09:47:45) 
[GCC 4.8.5 20150623 (Red Hat 4.8.5-39)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import chainer
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/kameyama/work/test/spack/opt/spack/linux-centos7-sandybridge/gcc-4.8.5/py-chainer-6.7.0-qzxc647vt63afqnckoyzlrc3e6cpt7v3/lib/python3.7/site-packages/chainer/__init__.py", line 10, in <module>
    from chainer import backends  # NOQA
  File "/home/kameyama/work/test/spack/opt/spack/linux-centos7-sandybridge/gcc-4.8.5/py-chainer-6.7.0-qzxc647vt63afqnckoyzlrc3e6cpt7v3/lib/python3.7/site-packages/chainer/backends/__init__.py", line 1, in <module>
    from chainer.backends import cuda  # NOQA
  File "/home/kameyama/work/test/spack/opt/spack/linux-centos7-sandybridge/gcc-4.8.5/py-chainer-6.7.0-qzxc647vt63afqnckoyzlrc3e6cpt7v3/lib/python3.7/site-packages/chainer/backends/cuda.py", line 37, in <module>
    import typing as tp  # NOQA
  File "/home/kameyama/work/test/spack/opt/spack/linux-centos7-sandybridge/gcc-4.8.5/py-typing-3.6.6-bnizzxvhfaapqifgtgj7aooalbjtenn7/lib/python3.7/site-packages/typing.py", line 1356, in <module>
    class Callable(extra=collections_abc.Callable, metaclass=CallableMeta):
  File "/home/kameyama/work/test/spack/opt/spack/linux-centos7-sandybridge/gcc-4.8.5/py-typing-3.6.6-bnizzxvhfaapqifgtgj7aooalbjtenn7/lib/python3.7/site-packages/typing.py", line 1004, in __new__
    self._abc_registry = extra._abc_registry
AttributeError: type object 'Callable' has no attribute '_abc_registry'
>>> 
</pre>
I add depends_on('^python@:3.6') for version 6.

Reply:
Hmm, in that case I guess we can keep the dependency how you had it. Whatever works. I don't know how many people will even be using the older version of Chainer anyway.

Reply:
I add old version to use for python2.